File: 226.翻转二叉树.c
/*
 * @lc app=leetcode.cn id=226 lang=c
 *
 * [226] 翻转二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void reverse(struct TreeNode *root)
{
  // 法一：前序遍历，递归
  // if (root == NULL)
  //   return;

  // struct TreeNode *temp;
  // temp = root->left;
  // root->left = root->right;
  // root->right = temp;

  // if (root->left != NULL)
  //   reverse(root->left);
  // if (root->right != NULL)
  //   reverse(root->right);

  // 法二：前序遍历，迭代
  // if (root == NULL)
  //   return;
  // int top;
  // struct TreeNode **stack;
  // stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 100);
  // top = -1;
  // stack[++top] = root;
  // while (top > -1)
  // {
  //   struct TreeNode *temp, *head;
  //   head = stack[top--];
  //   temp = head->left;
  //   head->left = head->right;
  //   head->right = temp;
  //   if (head->right)
  //     stack[++top] = head->right;
  //   if (head->left)
  //     stack[++top] = head->left;
  // }

  // 法三：前序遍历，通用法
  // if (root == NULL)
  //   return;
  // int top;
  // struct TreeNode **stack;
  // stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 100);
  // top = -1;
  // stack[++top] = root;
  // while (top > -1)
  // {
  //   struct TreeNode *node = stack[top--];
  //   if (node != NULL)
  //   {
  //     if (node->right)
  //       stack[++top] = node->right;
  //     if (node->left)
  //       stack[++top] = node->left;
  //     stack[++top] = node;
  //     stack[++top] = NULL;
  //   }
  //   else
  //   {
  //     struct TreeNode *temp;
  //     node = stack[top--];
  //     temp = node->left;
  //     node->left = node->right;
  //     node->right = temp;
  //   }
  // }
}

struct TreeNode *invertTree(struct TreeNode *root)
{
  reverse(root);
  return root;
}
// @lc code=end

File: store/226.翻转二叉树.js
/*
 * @lc app=leetcode.cn id=226 lang=javascript
 *
 * [226] 翻转二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
const preTraverse = (node) => {
  if (!node) return;
  let temp = node.left;
  node.left = node.right;
  node.right = temp;
  preTraverse(node.left);
  preTraverse(node.right);
};
const subTraverse = (node) => {
  if (!node) return;
  subTraverse(node.left);
  subTraverse(node.right);
  let temp = node.left;
  node.left = node.right;
  node.right = temp;
};
const midTraverse = (node) => {
  if (!node) return;
  midTraverse(node.left);
  let temp = node.left;
  node.left = node.right;
  node.right = temp;
  midTraverse(node.left);
};
var invertTree = function (root) {
  // preTraverse(root);
  // subTraverse(root);
  // midTraverse(root);
  // return root;

  const postTravel = (node) => {
    if (!node) {
      return null;
    }

    postTravel(node.left);
    const temp = node.left;
    node.left = node.right;
    node.right = temp;
    postTravel(node.left);
  };

  postTravel(root);

  return root;
};
// @lc code=end
